src/frameworks/core_foundation/time.rs

```diff
@@ -5,17 +5,94 @@
  */
 //! Time things including `CFAbsoluteTime`.
 
+use crate::abi::GuestArg;
 use crate::dyld::{export_c_func, FunctionExports};
+use crate::frameworks::core_foundation::CFTypeRef;
 use crate::frameworks::foundation::NSTimeInterval;
-use crate::objc::msg_class;
-use crate::Environment;
+use crate::libc::time::{time_t, timestamp_to_calendar_date};
+use crate::mem::SafeRead;
+use crate::objc::{msg_class, nil};
+use crate::{impl_GuestRet_for_large_struct, Environment};
+use std::time::SystemTime;
 
 pub type CFTimeInterval = NSTimeInterval;
 type CFAbsoluteTime = CFTimeInterval;
 
+#[derive(Copy, Clone, Debug, PartialEq)]
+#[repr(C, packed)]
+pub struct CFGregorianDate {
+    pub year: i32,    // SInt32
+    pub month: i8,    // SInt8
+    pub day: i8,      // SInt8
+    pub hours: i8,    // SInt8
+    pub minutes: i8,  // SInt8
+    pub seconds: f64, // double
+}
+unsafe impl SafeRead for CFGregorianDate {}
+impl_GuestRet_for_large_struct!(CFGregorianDate);
+impl GuestArg for CFGregorianDate {
+    const REG_COUNT: usize = 7;
+
+    fn from_regs(regs: &[u32]) -> Self {
+        CFGregorianDate {
+            year: GuestArg::from_regs(&regs[0..1]),
+            month: GuestArg::from_regs(&regs[1..2]),
+            day: GuestArg::from_regs(&regs[2..3]),
+            hours: GuestArg::from_regs(&regs[3..4]),
+            minutes: GuestArg::from_regs(&regs[4..5]),
+            seconds: GuestArg::from_regs(&regs[5..7]),
+        }
+    }
+    fn to_regs(self, regs: &mut [u32]) {
+        self.year.to_regs(&mut regs[0..1]);
+        self.month.to_regs(&mut regs[1..2]);
+        self.day.to_regs(&mut regs[2..3]);
+        self.hours.to_regs(&mut regs[3..4]);
+        self.minutes.to_regs(&mut regs[4..5]);
+        self.seconds.to_regs(&mut regs[5..7]);
```

Comment:
Wait, really? Did you verify this with Ghidra?

Reply:
Not really... How can I do that?

Reply:
Uh, well, you look at a function call involving the type and try to figure out the pattern of registers involved.

This grabbed my attention because the in-memory struct is seemingly quite compact (16 bytes), but if it's passed like this in registers or on the stack, then it's not so efficient (28 bytes). I'm not sure, but it might make sense if it has the same representation in both cases.

Reply:
> [if] from_regs and to_regs are consistent, it would work.

When passing data between touchHLE's own implementations, sure, but not when the struct is directly accessed by the app…

> I also confirmed that it works with small test app (see screenshot). Both cases (with 7 and with 4 registers) are working correctly.

Uhh… I'm scared now, I don't understand how you'd be able to write that C code and have it work for both 7 and 4 registers with the same results…

Reply:
Ok, i think i understand, my last test program has not involved any host function calling, that's why it's execution is independent of CFGregorianDate's GuestArg definition... Sorry for the noise.

Reply:
@hikari-no-yume  Ok, so... Finally I don't need to define GuestArg at all, only GuestRet :) So, I just removed GuestArg for CFGregorianDate from this PR.

Anyway, this is last version of test that I'm using:

```
int main(int argc, const char * argv[]) {
    time_t unixTime = time(NULL);
    CFAbsoluteTime at = CFAbsoluteTimeGetCurrent();
    printf("unix %d, abs %f, diff %d\n", unixTime, at, (int)ceil(unixTime - at));
    CFGregorianDate date = CFAbsoluteTimeGetGregorianDate(at, nil);
    printf("year %d, month %d, day %d, hour %d, minute %d, second %f\n",
        date.year, date.month, date.day, date.hour, date.minute, date.second);
    return 0;
}
```

Two things to notice:
1. Current CFAbsoluteTimeGetCurrent() returns time from startup of the programs, not a correct reference date. I haven't changed that.
2. seconds of CFGregorianDate do not have fractional part as i'm re-using timestamp_to_calendar_date() function.

Let me know if those need to be addressed before the merge.

<img width="1165" alt="Screenshot 2023-08-30 at 12 37 26" src="https://github.com/hikari-no-yume/touchHLE/assets/91161081/5b3596f5-140e-4b74-86de-41a96144a84f">

Reply:
Oh also, have you considered adding this test to TestApp? While we have to provide our own type definitions, if the date struct is declared with equivalent types to those in Apple's headers, it still serves as a useful test.

Reply:
> There's no complexity to extracting seconds, you can just call the .fract() method on the f64.

timestamp_to_calendar_date() returns tm struct with seconds defined as i32

> Oh also, have you considered adding this test to TestApp?

I have considered that, but given the current state of this PR this is cumbersome - I would need to recreate the PR from trunk with only comments which are not merged. I can do that, but this would screw up this PR history.

Reply:
> Uhh, since you're adding date conversion functions, I'd rather you use this opportunity to fix it, since otherwise all the dates are just nonsense. You can use Rust's SystemTime::now(), SystemTime::UNIX_EPOCH and the Duration functions that convert to/from floats. I suggest defining a constant like const APPLE_EPOCH: SystemTime = UNIX_EPOCH + …; with a precalculated offset.

Done.
(rust complained when i defined it as const because of non-const fn usage, so i defined it as a function finally.)

Reply:
> Done.  
> (rust complained when i defined it as const because of non-const fn usage, so i defined it as a function finally.)

Seems good to me, thanks. Nitpick: touchHLE only uses CamelCase functional names where Apple's APIs do. In this case, it's an internal function, so `snake_case` is preferred.

> > There's no complexity to extracting seconds, you can just call the .fract() method on the f64.

> timestamp_to_calendar_date() returns tm struct with seconds defined as i32

Oh, what I mean is you can bypass timestamp_to_calendar_date and extract the seconds directly from the float. I made a mistake though: `.frac()` would the fraction of a second, not the number of seconds in a minute, but you could add it to the result from that function.

Neither is a big issue, I can fix it myself on trunk if necessary.